lib/semantics/tools.cc

```diff
@@ -705,7 +705,10 @@ void ProcessParameterExpressions(
       if (paramValue != nullptr) {
         paramValue->SetExplicit(std::move(*expr));
       } else {
-        spec.AddParamValue(symbol->name(), ParamValue{std::move(*expr)});
+        ParamValue implicitParam{std::move(*expr)};
+        implicitParam.set_attr(details.attr());
+        spec.AddParamValue(
+            symbol->name(), ParamValue{std::move(implicitParam)});
```

Comment:
It would probably be worthwhile to add a constructor for `ParamValue` that takes an expr and `TypeParamAttr` (maybe defaulting to `Kind`). Then this would just be:
```
spec.AddParamValue(symbol->name(), ParamValue{std::move(*expr), details.attr()});
```
And you may decide you don't even need to add `GetLenParamValue()`.

Reply:
Thanks, I have actually made this the only kind of public constructor accepting expression so that the attribute cannot be forgotten to be safer. I removed `GetLenParamValue`.